sdk/python/lib/pulumi/x/automation/stack.py

```diff
@@ -0,0 +1,442 @@
+        """
+        Tries to create a new stack using the given workspace and stack name if the stack does not already exist,
+        or falls back to selecting the existing stack. If the stack does not exist,
+        it will be created and selected.
+
+        :param stack_name: The name identifying the Stack
+        :param workspace: The Workspace the Stack was created from.
+        :return: Stack
+        """
+        return Stack(stack_name, workspace, StackInitMode.CREATE_OR_SELECT)
+
+    def __init__(self, name: str, workspace: Workspace, mode: StackInitMode) -> None:
+        self.name = name
+        self.workspace = workspace
+
+        if mode is StackInitMode.CREATE:
+            workspace.create_stack(name)
+        elif mode is StackInitMode.SELECT:
+            workspace.select_stack(name)
+        elif mode is StackInitMode.CREATE_OR_SELECT:
+            try:
+                workspace.create_stack(name)
+            except StackAlreadyExistsError:
+                workspace.select_stack(name)
+
+    def up(self,
+           parallel: Optional[int] = None,
+           message: Optional[str] = None,
+           target: Optional[List[str]] = None,
+           expect_no_changes: Optional[bool] = None,
+           target_dependents: Optional[bool] = None,
+           replace: Optional[List[str]] = None,
+           on_output: Optional[OnOutput] = None,
+           program: Optional[PulumiFn] = None) -> UpResult:
+        """
+        Creates or updates the resources in a stack by executing the program in the Workspace.
+        https://www.pulumi.com/docs/reference/cli/pulumi_up/
+        """
+        program = program or self.workspace.program
+        extra_args = _parse_extra_args(**locals())
+        args = ["up", "--yes", "--skip-preview"]
+        args.extend(extra_args)
+
+        self.workspace.select_stack(self.name)
+        kind = ExecKind.LOCAL.value
+        on_exit = None
+
+        if program:
+            kind = ExecKind.INLINE.value
+            server = grpc.server(futures.ThreadPoolExecutor(max_workers=4),
+                                 options=_GRPC_CHANNEL_OPTIONS)
+            language_server = LanguageServer(program)
+            language_pb2_grpc.add_LanguageRuntimeServicer_to_server(language_server, server)
+
+            port = server.add_insecure_port(address="0.0.0.0:0")
+            server.start()
+
+            def on_exit():
+                language_server.on_pulumi_exit()
+                server.stop(0)
+            args.append(f"--client=127.0.0.1:{port}")
+
+        args.extend(["--exec-kind", kind])
+
+        try:
+            up_result = self._run_pulumi_cmd_sync(args, on_output)
+            outputs = self.outputs()
+            summary = self.info()
+            assert (summary is not None)
+            return UpResult(stdout=up_result.stdout, stderr=up_result.stderr, summary=summary, outputs=outputs)
+        except CommandError as exn:
```

Comment:
Are there other categories of error that could occur or are we guaranteed to get a command error? Maybe a finally block is the answer to this and my other questions about always running `on_exit` but not sure what the pythonic answer is.